### PR TITLE
[Rust] String-vs-str concept exercise

### LINKED ISSUE #1095: [Rust] Implement new Concept Exercise: string-vs-str
This issue describes how to implement the `string-vs-str` concept exercise for the rust track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

Please also watch the following video:

- [The Anatomy of a Concept Exercise](https://www.youtube.com/watch?v=gkbBqd7hPrA).

## Goal

Teach the student the differences between `String` and `&str` and how to work successfully with both.

## Learning objectives

- Know `String` is owned and mutable
- Know `&str` is shared and immutable
- Know `String` has `&str` methods because of shared traits 
- Know of performance characteristics

## Out of scope

- Machine level representation of either
- in depth performance differences

## Concepts

- mutability
- ownership
- slices

## Prerequisites

- basics
- slices
- str-use
- string-use

### Hints

- [string slices discussion in TRPL book][https://doc.rust-lang.org/book/ch04-03-slices.html#string-slices]
- [`&str` primitive documentation](https://doc.rust-lang.org/std/primitive.str.html)
- [`std::str`](https://doc.rust-lang.org/std/str/)
- [string types in RBE](https://doc.rust-lang.org/stable/rust-by-example/std/str.html)

### After

- [tutorial on string versus str in functions](http://hermanradtke.com/2015/05/03/string-vs-str-in-rust-functions.html)
- https://fasterthanli.me/blog/2020/working-with-strings-in-rust/
- https://doc.rust-lang.org/book/ch08-02-strings.html#storing-utf-8-encoded-text-with-strings

## Representer

- No changes needed

## Analyzer

- No changes needed

## Implementing

See our [implementation guide](https://github.com/exercism/v3/blob/master/languages/rust/howto/implementing-a-concept-exercise.md#how-to-implement-a-rust-concept-exercise) for guidance.
In addition you may want to make use of the shell script while in the concept directory, with the arg string-vs-str, e.g. `sh boil_exercise.sh string-vs-str`

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue.

### PR DESCRIPTION
**Pull request by gilescope**
Creating well escaped csv files is something every developer should strive for so I think this is a helpful exercise!
Also demostrates `consumption` of an owned value. (Fixes #1095 )

### CODE REVIEW

**Comment by gilescope** [comment]
It's says: "Forgot to run Prettier?" - is there a tool I should be running to check the .md files conform?

File: languages/rust/exercises/concept/string-vs-str/.meta/design.md
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/design.md:13]: Disagree with this assertion: the property "All strings are valid utf8" is essential to understand why indexing isn't implemented, for example.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/example.rs:1]:
```suggestion
pub struct RecordBuilder {
```

Nit: this should build a single record, not a full CSV file. Also, keeps it consistent with the instructions.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/example.rs:19]:
```suggestion
        if val.contains(",") || val.contains(r#"""#) || val.contains("\n") {
            self.content.push('"');
            self.content.push_str(&val.replace(r#"""#, r#""""#));
```

Nit: using raw literals reads a little more cleanly to me.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/example.rs:35]:
```suggestion
}

```

Files should end with a newline.

File: languages/rust/exercises/concept/string-vs-str/src/lib.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/src/lib.rs:25]:
```suggestion
}

```

Files should end with a newline.

File: languages/rust/exercises/concept/string-vs-str/tests/csv_builder.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/tests/csv_builder.rs:74]:
```suggestion
}

```

Files should end with a newline.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/example.rs:29]: Nit: Suggest removing this function and `test_two_line`. It doesn't teach anything new, and makes it accurate to call this a `RecordBuilder`.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/example.rs:31]:
```suggestion
    /// Consumes the builder and returns the CSV record.
```

File: languages/rust/exercises/concept/string-vs-str/tests/csv_builder.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/tests/csv_builder.rs:27]:
```suggestion
    assert_eq!(r#"ant,"ba""t",cat"#, &list);
```

Using raw strings makes this simpler to read. Adding a second double-quote within "bat" makes it more correct.

**Review by coriolinus — COMMENTED** [review]
Lots of small fixes. Overall seems positive. We might want to copy or move substantial documentation from `instructions.md` into `introduction.md`, the `String` concept, or some other relevant place. I'm not entirely up-to-date on the project layout.

Items marked "nit" are matters of opinion; if you disagree, feel free to just say so and mark them resolved.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (gilescope) [reply languages/rust/exercises/concept/string-vs-str/.meta/example.rs:29]: A `Record` is a bit of a nebulous term and probably means different things to different people. 
Most deevelopers would know what CSV is and how to construct it (if they're coming from other languages) and if not there's a spec in the RFC that makes it clear. For a lot of people CSV is the familiar part that they don't need to learn.

Comment (coriolinus) [reply languages/rust/exercises/concept/string-vs-str/.meta/example.rs:29]: `CSVRecordBuilder`, then? 

> A `Record` is a bit of a nebulous term.

Note the term's usage in [para 2.1 of RFC 4180](https://tools.ietf.org/html/rfc4180#page-2):

> Each record is located on a separate line, delimited by a line break (CRLF).

I feel justified in using it precisely according to the spec.

In any case, my main objection to the term `CSVBuilder` is just that you'd pretty much never want to actually build a full CSV file like this; even if you wanted to build the records manually instead of `#[derive(Serialize)]`, you'd stream them out to some `W: Write` as each was completed. If we're not going to do it for real, and it doesn't provide any education not present elsewhere in the file, why bother making the students implement the ability to handle multiple lines? And if the builder can't write multiple lines, it's not a general-purpose CSV builder, but a record builder.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/example.rs:28]: If we retain this function, which I'd prefer to avoid, then we'll need this edit to remain compliant with the spec:

```suggestion
        self.content.push("\r\n");
```

**Comment by efx** [comment]
Thanks for contributing @gilescope! I love the CSV idea. That is something I have to work with frequently so find this a grand combination of Rust and a practical problem.
> It's says: "Forgot to run Prettier?" - is there a tool I should be running to check the .md files conform?
Pro tip: you can run the formatting script locally if you want with `EXERCISM_PRETTIER_VERSION=2.1.2 ./bin/format.sh`. It runs it on everything in the repository.

File: languages/rust/exercises/concept/string-vs-str/.meta/design.md
Comment (efx) [inline languages/rust/exercises/concept/string-vs-str/.meta/design.md:33]:
```suggestion
    (string slices discussion in TRPL book)[https://doc.rust-lang.org/book/ch04-03-slices.html#string-slices]
```

Comment (coriolinus) [reply languages/rust/exercises/concept/string-vs-str/.meta/design.md:33]: I think you replaced the wrong set of brackets with parentheses.

It happens to everyone.

Comment (efx) [inline languages/rust/exercises/concept/string-vs-str/.meta/design.md:33]: nitpick: could we make these links consistently formatted? 
```md
[string slices discussion in TRPL book](https://doc.rust-lang.org/book/ch04-03-slices.html#string-slices)
[&str primitive documentation](https://doc.rust-lang.org/std/primitive.str.html)
[std::str](https://doc.rust-lang.org/std/str/)
[string types in RBE](https://doc.rust-lang.org/stable/rust-by-example/std/str.html)
```
should do it.
Thanks @coriolinus for noting my earlier incorrect suggestion. Didn't mean to delete your comment FWIW; GitHub seems to do that for deleting the original suggestion.

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (gilescope) [reply languages/rust/exercises/concept/string-vs-str/.meta/example.rs:29]: CSVRecordBuilder I can get behind. `writeln!(self.content, "");` would do the trick cross platform. I do accept that doing a cross platform line ending is hard in rust. I asked recently on the internals and they said basically that `\r` was dead and that even notepad understood just `\n` these days... 

Maybe we just implement that function for them to show how its done.

Comment (gilescope) [reply languages/rust/exercises/concept/string-vs-str/.meta/example.rs:29]: Ah your right let's keep it simple - I'll just say in the instructions that we're focusing on a single line.

**Comment by gilescope** [comment]
Ok updated version.

There's no slicing done at all. I think that's fine and lives in a more &str focused exercise. - maybe something parsing related? (A get out of this without allocating exercise... -- hmm makes me wonder if we can have some exercises that are no_std so that we can teach &str without there being the possibiliity of allocating... yes there's some fun potentially to be had there.)

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:48]:
```suggestion
(The [heapless](https://crates.io/crates/heapless) crate is often used when manipulating
```

Markdown links are always brackets first, parens second.

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:43]:
```suggestion
- `let s: String = "my str".into();`
```

Intent is primarily to get rid of the spacing before the colon. I don't think that this example is unexplicit, though: it clearly shows what the output type is.

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:36]:
```suggestion
`String` implements `Deref<Target=str>`. This trait has some compiler special casing: it means that any reference to a `String` can automatically, transparently be coerced into an `&str`. This implies the following:

- Any code which expects, for example, an `&str` argument can accept an `&String` instead.
- Any method implemented for `&str` can be called on a `String` as well. 
```

Precision: it's factually wrong to say that "`&String::new()` is of type `&str`", but it's still worth digging into the magic of `Deref`.

File: languages/rust/exercises/concept/string-vs-str/.meta/design.md
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.meta/design.md:40]: Is this line a TODO, a label for the following link, or something else? It should be clarified or removed; as is, it's not very helpful.

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (gilescope) [reply languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:36]: ah your right it's a `&String` type but like `&str` the both deref into `str`.

It makes me think we should mention in passing `AsRef<str>` even though generics aren't yet covered.

File: languages/rust/exercises/concept/string-vs-str/.meta/design.md
Comment (gilescope) [reply languages/rust/exercises/concept/string-vs-str/.meta/design.md]: sorry link was missing. Have added hints.md also.

File: languages/rust/exercises/concept/string-vs-str/.docs/hints.md
Comment (coriolinus) [inline languages/rust/exercises/concept/string-vs-str/.docs/hints.md:10]: Nit: I'd omit this link; it's both quite old and makes some factually incorrect assertions, i.e. " String is just a vector of 8-byte chars". It's not a big enough deal to hold up matters on this PR, though.

**Review by coriolinus — APPROVED** [review]
Thanks for all the work @gilescope!

As far as Rust goes, this is mergeable. However, I have not checked against the v3 format or style guide. In particular, I think there should be some kind of json file declaraing the prerequisites, concepts unlocked, etc. I believe @ErikSchierboom is the subject expert there.

**Comment by gilescope** [comment]
Cool - I am sure more things will come to me once I see it in situ on the site. I'm sure we will revisit as I know more about the process.

File: languages/rust/exercises/concept/string-vs-str/src/lib.rs
Comment (ErikSchierboom) [inline languages/rust/exercises/concept/string-vs-str/src/lib.rs:8]: To help the student figure out what to do, we often pass in an argument to let the student know what to implement:

```suggestion
        unimplemented!("Please implement the `CSVRecordBuilder.new` function")
```

or something like that. See [this Rust exercise](https://github.com/exercism/v3/blob/master/languages/rust/exercises/concept/tuples/src/lib.rs#L2) for something similar.

File: languages/rust/exercises/concept/string-vs-str/src/lib.rs
Comment (ErikSchierboom) [inline languages/rust/exercises/concept/string-vs-str/src/lib.rs:13]: See above comment.

File: languages/rust/exercises/concept/string-vs-str/src/lib.rs
Comment (ErikSchierboom) [inline languages/rust/exercises/concept/string-vs-str/src/lib.rs:18]: See above comment.

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (ErikSchierboom) [inline languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:1]: This file is well-written! It is slightly on the longer side though. As per [the spec](https://github.com/exercism/v3-docs/blob/master/anatomy/tracks/concept-exercises.md#file-docsintroductionmd):

> The information provided should give the student just enough context to figure out the solution themselves.

What this means is that this file contain _just enough_ information for the student to figure things out by themselves. There are often aspects of a concept that are very interesting, but that the student does not need to understand in order to solve the exercise. This type of information should be removed from the `introduction.md` file. There is another place to put this information though, which is the concept's `about.md` file (see [the spec](https://github.com/exercism/v3-docs/blob/master/anatomy/tracks/concepts.md#file-aboutmd)). This file is shown once the student finishes the exercise, and can go in far more detail on a concept than the `introduction.md` allows for.

I'll leave it up to you to decide what is essential for the student :) A good way to go about this would be to check the example solution, as that should be solution you expect the student to write.

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (ErikSchierboom) [inline languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:8]: What does the `invalid` do? Never seen that used before, interesting!

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (ErikSchierboom) [inline languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:1]: > (In Java or C#, this type is called `StringBuilder`.)

This bit will only really be useful to Java or C# programmers, which means that it won't benefit students not familiar with those languages. It can be very useful information, so you could:

1. Leave it as is
2. Rewrite it slightly, to something like: ```In some languages, this type is referred to as a `StringBuilder`.```

Whatever you prefer :)

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (ErikSchierboom) [inline languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:55]: As per [the spec](https://github.com/exercism/v3-docs/blob/master/anatomy/tracks/concept-exercises.md#file-docsintroductionmd):

> Links should be used sparingly, if at all. While a link explaining a complex topic like recursion might be useful, for most concepts the links will provide more information than needed so explaining things concisely inline should be the aim.

This section sounds very interesting, but it looks like the student does not need to know this to solve the exercise. This is perfect for the concept's `about.md` file though.

**Review by ErikSchierboom — APPROVED** [review]
This is great work! I've left some comments. Feel free to ask any questions if you have them. Some general comments on this PR that don't apply to a single file:

- The exercise is currently named `string-vs-str`. We switched to having exercise being named after their story/theme (see [this document](https://github.com/exercism/v3/blob/master/docs/maintainers/determining-concepts.md#naming-concept-exercises)) a while back. Could you rename the exercise and update its directory name accordingly?

- There are some missing files. As @coriolinus mentioned, each track has a `config.json` file that contains the track's metadata. This metadata includes the Concept Exercises and the Concepts. Could you add metadata for this Concept Exercises and its concept(s) to [Rust's `config.json` file](https://github.com/exercism/v3/blob/master/languages/rust/config.json)?

The concept-specific files are missing. See [the spec](https://github.com/exercism/v3-docs/blob/master/anatomy/tracks/concepts.md) for more information.

I've approved this PR so you can decide to either merge it now and do follow-up PRs to address the comments, or keep this PR open and address the comments in this PR. Whatever you prefer! Great job!

File: languages/rust/exercises/concept/string-vs-str/.meta/example.rs
Comment (coriolinus) [reply languages/rust/exercises/concept/string-vs-str/.meta/example.rs:17]: That's not a regular expression; that's raw string literal syntax. It's used in a few places here to denote double-quotes without escapes all over the place. We could maybe add a small syntax note in the docs somewhere.

File: languages/rust/exercises/concept/string-vs-str/.docs/introduction.md
Comment (coriolinus) [reply languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:8]: The only place I'm really familiar with seeing multiple tags on a code fence like that is when writing Rust documentation; tagging a snippet as "invalid" will render it with a red box surrounding it and a popup noting that this is invalid. It also won't attempt to compile it as a doctest. 

That said, I did check; the GFM renderer appears to properly render the sample as Rust, which is sufficient.

Comment (ErikSchierboom) [reply languages/rust/exercises/concept/string-vs-str/.docs/introduction.md:8]: Thanks for the explanation!

**Pull request merged**
